File: 0133-clone-graph/0133-clone-graph.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var neighbors: [Node?]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.neighbors = []
 *     }
 * }
 */

class Solution {
    func cloneGraph(_ node: Node?) -> Node? {
        var oldToNew: [Node: Node] = [:]

        func clone(_ node: Node?) -> Node? {
            guard let node = node else { return nil }

            if let newNode = oldToNew[node] {
                return newNode
            }
            var copy = Node(node.val)
            oldToNew[node] = copy
            for neighbour in node.neighbors {
                copy.neighbors.append(clone(neighbour))
            }

            return copy
        }

        return clone(node)
    }
}

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        """
        whenever we add to visit, we need to create a new node
        if we see something in visited, add the edge to neighbours but don't dfs into it
        Note: for DFS, the stack is the recursion itself *mind-blown

        Time: O(V + E), O(V) since E will be at most V
        Space: O(V)
        """
        self.visited = set()
        self.valToNode = {}
        def dfs(node):
            if not node:
                return None

            newNode = Node(node.val)
            self.visited.add(node.val)
            self.valToNode[node.val] = newNode
            for neighbour in node.neighbors:
                if neighbour.val not in self.visited:
                    newNode.neighbors.append(dfs(neighbour))
                else:
                    newNode.neighbors.append(self.valToNode[neighbour.val])
            return newNode

        return dfs(node)

